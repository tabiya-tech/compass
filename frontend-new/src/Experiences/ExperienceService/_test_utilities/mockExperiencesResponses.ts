import { Experience, Skill, WorkType } from "src/Experiences/ExperienceService/Experiences.types";
import { faker } from "@faker-js/faker";
import { v4 as uuidv4 } from "uuid";
import { format } from "date-fns";

export const mockExperiences: Experience[] = [
  {
    UUID: "c3c8b43d-73dd-4c6d-9e71-010492e86d5e",
    start_date: "2022",
    end_date: "Present",
    experience_title: "Project Manager",
    company: "Business Inc",
    location: "New York, NY",
    work_type: WorkType.FORMAL_SECTOR_WAGED_EMPLOYMENT,
    top_skills: [
      {
        UUID: "2512652c-49ba-4751-b654-8fc525ce98ca",
        preferredLabel: "Management",
        description: "The process of dealing with or controlling things or people",
        altLabels: ["Leadership", "Administration"],
      },
      {
        UUID: "815f3660-4c7f-43ba-859d-cf50dd527fe0",
        preferredLabel: "Communication",
        description: "The imparting or exchanging of information or news",
        altLabels: ["Interpersonal Communication", "Public Speaking"],
      },
    ],
  },
  {
    UUID: "c3c8b43d-73dd-4c6d-9e71-010492e86d5e",
    start_date: "2022",
    end_date: "Present",
    experience_title: "Project Manager",
    company: "Business Inc",
    location: "New York, NY",
    work_type: WorkType.FORMAL_SECTOR_UNPAID_TRAINEE_WORK,
    top_skills: [
      {
        UUID: "2512652c-49ba-4751-b654-8fc525ce98ca",
        preferredLabel: "Management",
        description: "The process of dealing with or controlling things or people",
        altLabels: ["Leadership", "Administration"],
      },
      {
        UUID: "815f3660-4c7f-43ba-859d-cf50dd527fe0",
        preferredLabel: "Communication",
        description: "The imparting or exchanging of information or news",
        altLabels: ["Interpersonal Communication", "Public Speaking"],
      },
    ],
  },
];

const allWorkTypes = [
  WorkType.SELF_EMPLOYMENT,
  WorkType.FORMAL_SECTOR_WAGED_EMPLOYMENT,
  WorkType.FORMAL_SECTOR_UNPAID_TRAINEE_WORK,
  WorkType.UNSEEN_UNPAID,
];

const skillsLabels = [
  "work in an organised manner",
  "administer ingredients in food production",
  "sales activities",
  "unload supplies",
  "perform outdoor cleaning activities",
  "selling Kotas",
  "helping Brother with Car",
  "community Volunteering",
  "ensured the project was completed on time",
  "took care of the garden",
  "communication",
  "helped with the family business",
  "time management",
  "organise product display",
  "assist customers",
  "maintain work area cleanliness",
  "teamwork",
  "attention to details",
  "problem solving",
];

const generateRandomSkill = (usedLabels: string[]): Skill => {
  let randomLabel;
  do {
    randomLabel = skillsLabels[Math.floor(Math.random() * skillsLabels.length)];
  } while (usedLabels.includes(randomLabel));
  usedLabels.push(randomLabel);
  return {
    UUID: uuidv4(),
    preferredLabel: randomLabel,
    description: faker.hacker.phrase(),
    altLabels: [faker.hacker.verb(), faker.hacker.adjective()],
  };
};

const formatShortDate = (date: Date): string => format(date, "MMM yyyy");

const generateRandomDate = (): string => {
  return formatShortDate(faker.date.past());
};

const generateStandaloneDate = (): string => {
  const chance = Math.random();
  return chance < 0.5 ? "A long time ago" : "Since I was five";
};

const generateRandomExperience = (workType?: WorkType): Experience => {
  const useStandaloneDate = Math.random() < 0.5;
  const startDate = useStandaloneDate ? generateStandaloneDate() : generateRandomDate();
  const endDate = useStandaloneDate ? "" : Math.random() > 0.5 ? generateRandomDate() : "Present";
  const randomWorkType = workType || allWorkTypes[Math.floor(Math.random() * allWorkTypes.length)];
  const usedLabels: string[] = [];

  return {
    UUID: uuidv4(),
    start_date: startDate,
    end_date: endDate,
    experience_title: faker.person.jobTitle(),
    company: faker.company.name(),
    location: faker.location.city(),
    work_type: randomWorkType,
    top_skills: [
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
      generateRandomSkill(usedLabels),
    ],
  };
};

export const generateRandomExperiences = (count: number): Experience[] => {
  const experiences: Experience[] = [];

  // ensure at least one experience is present for each work type
  allWorkTypes.forEach((workType) => {
    experiences.push(generateRandomExperience(workType));
  });

  // generate the rest of the experiences
  for (let i = 0; i < count; i++) {
    experiences.push(generateRandomExperience());
  }
  return experiences;
};
