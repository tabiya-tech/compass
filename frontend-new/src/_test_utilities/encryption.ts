import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
