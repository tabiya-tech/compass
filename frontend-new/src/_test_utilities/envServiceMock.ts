jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
