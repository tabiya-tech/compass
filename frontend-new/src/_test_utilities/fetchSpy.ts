import { ExtendedRequestInit } from "src/utils/customFetch/customFetch";

export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}

// spy on authenticated fetch requests
export function setupAPIServiceSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });

  jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
  return jest.spyOn(require("src/utils/customFetch/customFetch"), "customFetch");
}

export function expectCorrectFetchRequest(fetchSpy: jest.SpyInstance, expectedUrl: string, expectedConfig: ExtendedRequestInit) {
  expect(fetchSpy).toHaveBeenCalledTimes(1);
  const [actualUrl, actualConfig] = fetchSpy.mock.calls[0];
  expect(actualUrl).toBe(expectedUrl);
  expect(actualConfig.method).toBe(expectedConfig.method);

  // Compare headers if they exist
  if(actualConfig.headers) {
    // Handle headers comparison - convert Headers object to plain object if needed
    const actualHeaders = actualConfig.headers instanceof Headers
      ? Object.fromEntries(actualConfig.headers.entries())
      : actualConfig.headers;
    expect(actualHeaders).toEqual(
      expect.objectContaining(expectedConfig.headers),
    );
  }

  // Compare bodies by parsing JSON if they exist
  if (actualConfig.body && expectedConfig.body) {
    let actualBody: unknown;
    let expectedBody: unknown;
    
    try {
      actualBody = JSON.parse(JSON.stringify(actualConfig.body));
      expectedBody = JSON.parse(JSON.stringify(expectedConfig.body));
    } catch (e) {
      console.error('Error parsing request bodies:', e);
      throw new Error('Failed to parse request bodies for comparison');
    }
    expect(actualBody).toEqual(expectedBody);
  }

  // Check any extra properties if provided
  type ExtendedConfigKey = keyof ExtendedRequestInit;
  const extraProperties: ExtendedConfigKey[] = [
    'expectedStatusCode',
    'serviceName',
    'serviceFunction',
    'failureMessage',
    'expectedContentType'
  ] as ExtendedConfigKey[];

  extraProperties.forEach((key) => {
    if (expectedConfig[key] !== undefined) {
      expect(actualConfig[key]).toEqual(expectedConfig[key]);
    }
  });
}

