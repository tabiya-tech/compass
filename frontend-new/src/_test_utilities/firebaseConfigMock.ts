import { AuthProviderIds } from "src/auth/auth.types";

const auth = jest.fn(() => ({
  signInWithCustomToken: jest.fn(),
  signInWithEmailAndPassword: jest.fn(),
  createUserWithEmailAndPassword: jest.fn(),
  signOut: jest.fn(),
  GoogleAuthProvider: { PROVIDER_ID: AuthProviderIds.GOOGLE },
}));

export function mockFirebaseConfig() {
  jest.spyOn(require("src/auth/firebaseConfig"), "auth").mockImplementation(auth);
}
