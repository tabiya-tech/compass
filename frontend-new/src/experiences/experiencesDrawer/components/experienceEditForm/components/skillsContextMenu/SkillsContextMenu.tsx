import { Box, Divider, ListItemText, Menu, MenuItem, Typography, useMediaQuery, useTheme } from "@mui/material";
import { MenuItemConfig } from "src/theme/ContextMenu/menuItemConfig.types";
import React from "react";

export interface SkillsContextMenuProps {
  items: MenuItemConfig[];
  open: boolean;
  anchorEl: HTMLElement | null;
  notifyOnClose: () => void;
}

const uniqueId = "54bb15e1-a83f-49f8-84e1-8158ef7b2ec1";
export const DATA_TEST_ID = {
  SKILL_MENU: `${uniqueId}-skill-menu`,
  SKILL_MENU_ITEM: `${uniqueId}-skill-menu-item`,
  SKILL_MENU_ITEM_TEXT: `${uniqueId}-skill-menu-item-text`,
  SKILL_MENU_ITEM_DESCRIPTION: `${uniqueId}-skill-menu-item-description`,
  SKILL_MENU_HEADER_MESSAGE: `${uniqueId}-skill-menu-header-message`,
};

const SkillsContextMenu: React.FC<Readonly<SkillsContextMenuProps>> = (props) => {
  const theme = useTheme();
  const isSmallMobile = useMediaQuery(theme.breakpoints.down("sm"));
  const isMobile = useMediaQuery(theme.breakpoints.down("md"));
  const handleItemClick = (item: MenuItemConfig) => {
    props.notifyOnClose();
    item.action();
  };

  return (
    <Menu
      elevation={2}
      anchorOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      transformOrigin={{
        vertical: "bottom",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
      data-testid={DATA_TEST_ID.SKILL_MENU}
      slotProps={{
        paper: {
          sx: {
            width: isMobile ? "100%" : "60%",
            overflow: "hidden",
          },
        },
      }}
      MenuListProps={{
        sx: {
          paddingTop: 0,
        },
      }}
    >
      <Box
        position="sticky"
        top={0}
        bgcolor={theme.palette.background.paper}
        paddingX={theme.fixedSpacing(theme.tabiyaSpacing.md)}
        paddingTop={theme.fixedSpacing(theme.tabiyaSpacing.md)}
        width="100%"
        zIndex={1}
      >
        <Typography
          variant="caption"
          fontWeight="bold"
          color={theme.palette.text.secondary}
          data-testid={DATA_TEST_ID.SKILL_MENU_HEADER_MESSAGE}
        >
          These skills were identified by Compass but are considered less relevant to your experience
        </Typography>
        <Divider sx={{ paddingTop: theme.fixedSpacing(theme.tabiyaSpacing.sm) }} />
      </Box>
      <Box
        sx={{
          maxHeight: isSmallMobile ? "80vh" : "60vh",
          overflowY: "scroll",
          overflowX: "hidden",
          "&::-webkit-scrollbar": {
            width: theme.fixedSpacing(theme.tabiyaSpacing.sm),
          },
          "&::-webkit-scrollbar-thumb": {
            backgroundColor: theme.palette.primary.main,
            borderRadius: theme.fixedSpacing(theme.tabiyaSpacing.xs),
          },
          paddingTop: theme.fixedSpacing(theme.tabiyaSpacing.sm),
        }}
      >
        {props.items.map((item) => (
          <MenuItem
            onClick={() => handleItemClick(item)}
            data-testid={DATA_TEST_ID.SKILL_MENU_ITEM}
            disabled={item.disabled}
            key={item.id}
          >
            <ListItemText>
              <Typography
                whiteSpace="normal"
                variant="caption"
                color={item.textColor ?? "secondary"}
                data-testid={DATA_TEST_ID.SKILL_MENU_ITEM_TEXT}
              >
                {item.text}
              </Typography>
              <Typography
                variant="caption"
                sx={{
                  display: "-webkit-box",
                  WebkitLineClamp: 3,
                  WebkitBoxOrient: "vertical",
                  overflow: "hidden",
                  width: "100%",
                  textOverflow: "ellipsis",
                  whiteSpace: "normal",
                }}
                data-testid={DATA_TEST_ID.SKILL_MENU_ITEM_DESCRIPTION}
              >
                {item.description}
              </Typography>
            </ListItemText>
          </MenuItem>
        ))}
      </Box>
    </Menu>
  );
};

export default SkillsContextMenu;
