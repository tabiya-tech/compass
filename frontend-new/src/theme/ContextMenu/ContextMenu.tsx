import { Box, Divider, Icon, ListItemIcon, ListItemText, Menu, MenuItem, Typography, useTheme } from "@mui/material";
import { MenuItemConfig } from "./menuItemConfig.types";

export interface ContextMenuProps {
  items: MenuItemConfig[];
  open: boolean;
  anchorEl: HTMLElement | null;
  notifyOnClose: () => void;
  headerMessage?: string;
}

const uniqueId = "b7499b01-8082-4209-8667-c7d559a70caf";
export const DATA_TEST_ID = {
  MENU: `${uniqueId}-menu`,
  MENU_ITEM: `${uniqueId}-menu-item`,
  MENU_ITEM_ICON: `${uniqueId}-menu-item-icon`,
  MENU_ITEM_TEXT: `${uniqueId}-menu-item-text`,
  MENU_HEADER_MESSAGE: `${uniqueId}-menu-header-message`,
};

function ContextMenu(props: Readonly<ContextMenuProps>) {
  const theme = useTheme();
  const handleItemClick = (item: MenuItemConfig) => {
    props.notifyOnClose();
    item.action();
  };

  return (
    <Menu
      elevation={2}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
      data-testid={DATA_TEST_ID.MENU}
    >
      {props.headerMessage && (
        <Box>
          <Box
            sx={{
              maxWidth: 250,
              px: theme.fixedSpacing(theme.tabiyaSpacing.md),
              py: theme.fixedSpacing(theme.tabiyaSpacing.sm),
            }}
            data-testid={DATA_TEST_ID.MENU_HEADER_MESSAGE}
          >
            <Typography variant="caption" fontWeight="bold" color={theme.palette.text.secondary}>
              {props.headerMessage}
            </Typography>
          </Box>
          <Divider />
        </Box>
      )}
      {props.items.map((item) => (
        <MenuItem
          onClick={() => handleItemClick(item)}
          data-testid={DATA_TEST_ID.MENU_ITEM}
          disabled={item.disabled}
          key={item.id}
        >
          {item.icon && (
            <ListItemIcon data-testid={DATA_TEST_ID.MENU_ITEM_ICON}>
              <Icon sx={{ color: theme.palette.text.secondary }}>{item.icon}</Icon>
            </ListItemIcon>
          )}
          <ListItemText data-testid={DATA_TEST_ID.MENU_ITEM_TEXT}>
            <Typography variant="caption" color={item.textColor ?? "secondary"}>
              {item.text}
            </Typography>
            {item.description && (
              <Typography variant="caption" display="flex" whiteSpace="normal">
                {item.description}
              </Typography>
            )}
          </ListItemText>
        </MenuItem>
      ))}
    </Menu>
  );
}

export default ContextMenu;
