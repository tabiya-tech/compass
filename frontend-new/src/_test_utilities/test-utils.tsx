// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
