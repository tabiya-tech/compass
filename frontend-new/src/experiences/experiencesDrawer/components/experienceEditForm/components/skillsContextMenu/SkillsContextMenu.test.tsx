// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import { ContextMenuProps } from "src/theme/ContextMenu/ContextMenu";
import SkillsContextMenu, { DATA_TEST_ID } from "./SkillsContextMenu";
import { within } from "@testing-library/react";

describe("SkillsContextMenu", () => {
  const stdGivenProps: ContextMenuProps = {
    anchorEl: document.createElement("div"), // Mock an HTMLElement
    open: true,
    notifyOnClose: jest.fn(),
    items: [],
  };

  test("should render correctly the skills menu open", () => {
    // GIVEN the following menu items
    const givenItems = [
      {
        id: "1",
        text: "foo",
        description: "Description for foo",
        action: jest.fn(),
        disabled: false,
      },
      {
        id: "2",
        text: "bar",
        description: "Description for bar",
        action: jest.fn(),
        disabled: false,
      },
    ];

    // WHEN the component is rendered with the given items
    render(<SkillsContextMenu {...stdGivenProps} items={givenItems} />);

    // THEN expect the SkillsContextMenu to be visible
    const actualMenu = screen.getByTestId(DATA_TEST_ID.SKILL_MENU);
    expect(actualMenu).toBeInTheDocument();
    // AND the skill menu header message to be visible
    const headerMessage = screen.getByTestId(DATA_TEST_ID.SKILL_MENU_HEADER_MESSAGE);
    expect(headerMessage).toBeInTheDocument();
    // AND the menu items to be visible
    const actualMenuItems = screen.getAllByTestId(DATA_TEST_ID.SKILL_MENU_ITEM);
    expect(actualMenuItems).toHaveLength(givenItems.length);
    // AND every menu item to have the correct text and description
    givenItems.forEach((item, index) => {
      const menuItem = within(actualMenuItems[index]);
      expect(menuItem.getByTestId(DATA_TEST_ID.SKILL_MENU_ITEM_TEXT)).toHaveTextContent(item.text);
      expect(menuItem.getByTestId(DATA_TEST_ID.SKILL_MENU_ITEM_DESCRIPTION)).toHaveTextContent(item.description);
    });
    // AND to match the snapshot
    expect(actualMenu).toMatchSnapshot();
    // AND no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
  });

  test("should render correctly the skills menu closed", () => {
    // GIVEN the skill menu is closed
    const givenProps = { ...stdGivenProps, open: false };

    // WHEN the component is rendered with the given items but closed
    render(<SkillsContextMenu {...givenProps} />);

    // THEN expect the SkillsContextMenu to not be visible
    const actualMenu = screen.queryByTestId(DATA_TEST_ID.SKILL_MENU);
    expect(actualMenu).not.toBeInTheDocument();
  });

  test("should call the action of a menu item when clicked and the notifyOnClose function", () => {
    // GIVEN the following menu items
    const givenItems = [
      {
        id: "1",
        text: "foo",
        description: "Description for foo",
        action: jest.fn(),
        disabled: false,
      },
    ];

    // WHEN the component is rendered with the given items
    render(<SkillsContextMenu {...stdGivenProps} items={givenItems} />);

    // AND the first menu item is clicked
    const actualMenuItem = screen.getByTestId(DATA_TEST_ID.SKILL_MENU_ITEM);
    actualMenuItem.click();

    // THEN expect the action of the first menu item to be called
    expect(givenItems[0].action).toHaveBeenCalled();
    // AND expect notifyOnClose to be called
    expect(stdGivenProps.notifyOnClose).toHaveBeenCalled();
  });
});
