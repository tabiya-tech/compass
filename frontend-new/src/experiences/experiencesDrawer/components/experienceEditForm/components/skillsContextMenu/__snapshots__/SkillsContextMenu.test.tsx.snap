// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsContextMenu should render correctly the skills menu open 1`] = `
<div
  class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
  data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiPopover-paper MuiMenu-paper MuiMenu-paper css-va7vk9-MuiPaper-root-MuiPopover-paper-MuiMenu-paper"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root MuiList-padding MuiMenu-list css-erpndv-MuiList-root-MuiMenu-list"
      role="menu"
      tabindex="-1"
    >
      <div
        class="MuiBox-root css-1dyflr6"
        tabindex="0"
      >
        <span
          class="MuiTypography-root MuiTypography-caption css-u85l1t-MuiTypography-root"
          data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-header-message"
        >
          These skills were identified by Compass but are considered less relevant to your experience
        </span>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth css-7xd7gm-MuiDivider-root"
        />
      </div>
      <div
        class="MuiBox-root css-io668w"
      >
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-f908r0-MuiButtonBase-root-MuiMenuItem-root"
          data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1bcerjy-MuiTypography-root"
            >
              <span
                class="MuiTypography-root MuiTypography-caption css-tp85g4-MuiTypography-root"
                data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item-text"
              >
                foo
              </span>
              <span
                class="MuiTypography-root MuiTypography-caption css-n68kj6-MuiTypography-root"
                data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item-description"
              >
                Description for foo
              </span>
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
        <li
          class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters MuiMenuItem-root MuiMenuItem-gutters css-f908r0-MuiButtonBase-root-MuiMenuItem-root"
          data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item"
          role="menuitem"
          tabindex="-1"
        >
          <div
            class="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1bcerjy-MuiTypography-root"
            >
              <span
                class="MuiTypography-root MuiTypography-caption css-tp85g4-MuiTypography-root"
                data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item-text"
              >
                bar
              </span>
              <span
                class="MuiTypography-root MuiTypography-caption css-n68kj6-MuiTypography-root"
                data-testid="54bb15e1-a83f-49f8-84e1-8158ef7b2ec1-skill-menu-item-description"
              >
                Description for bar
              </span>
            </span>
          </div>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </li>
      </div>
    </ul>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
